### ISSUE TITLE
How to follow best practice with this package (i.e. not using username/password) ?

### ISSUE THREAD

**Issue by activebiz** [oblador/react-native-keychain#291]
Hi,

I am new to this package so apologies in advance if this sounds basic question. I understand the following statement which is on the readme.md. 

```Follow best practices and do not store user credentials on a device. Instead use tokens or other forms of authentication and re-ask for user credentials before performing sensitive operations.```

However all the methods in the documentation and in he keychainexample is using username/passwords. So I am not sure how do I use tokens to store the sensitive data to the device using this library?!

Thanks

**Comment by Brqqq** [comment]
The username/password examples are just to make the API usage clear I think.

What you need to store depends on your particular situation. Think of it like this: what can you store so that if there is a security breach, the impact will be minimal? 

Token-based approach is not about this library, but about your login system. If you use oauth2 on your service/website for example, you will likely get some access token after logging in. This access token is sent with each API call to show the server you're authenticated. It can also be revoked. So you should store the access token rather than username/password. 

In the event of a security breach, you can simply revoke the token and there will be no further damage. If the attacker gets a username/password, the impact is very big. They can for example try this username/password on other services. Storing passwords should not be taken lightly.

The other question to ask is: should you be storing this info at all? Why do you want to store a password? Wouldn't it be safer if the user just has to re-type their password? Maybe you can avoid the whole issue in the first place.

**Comment by louiszawadzki** [comment]
To add to @Brqqq's answer, if you want to authenticate users on your server and tokens is not a solution (for whatever reason), maybe you should consider following what react-native-biometrics recommends: https://github.com/SelfLender/react-native-biometrics#usage.

> When a user enrolls in biometrics, a key pair is generated. The private key is stored securely on the device and the public key is sent to a server for registration. When the user wishes to authenticate, the user is prompted for biometrics, which unlocks the securely stored private key. Then a cryptographic signature is generated and sent to the server for verification. The server then verifies the signature. If the verification was successful, the server returns an appropriate response and authorizes the user.

See also https://android-developers.googleblog.com/2015/10/new-in-android-samples-authenticating.html

**Comment by paulosborne** [comment]
> So you should store the access token rather than username/password.

I think the OP is actually asking what is the best practice for securely storing arbitrary data (such as a token) using this library. The API seems to be heavily focused on storing username and passwords. For example, the method `setGenericPassword` has the following signature

```javascript
setGenericPassword(username, password, [{ accessControl, accessible, accessGroup, service, securityLevel }])
```
That seems more than just an example to make usage clearer, I would expect something more like 

```javascript
setItem(key, value, options)
getItem(key)
```

**Comment by Aung-Myint-Thein** [comment]
I have 2 values (tokens) that i would like to store but I am not sure what is the best way to store them.. any recommendations?

**Comment by Iltimirov** [comment]
I found two ways how it can be implemented

interface
```typescript
type SetSecureValue = (
  key: string,
  value: string,
  accessible?: Keychain.ACCESSIBLE,
) => Promise<void>
type GetSecureValue = (key: string) => Promise<string | false>
type RemoveSecureValue = (key: string) => Promise<void>
```

1. with generic password (use optional service param)
```typescript
export const setSecureValue: SetSecureValue = (
  key,
  value,
  accessible = Keychain.ACCESSIBLE.ALWAYS_THIS_DEVICE_ONLY,
) => Keychain.setGenericPassword(key /* <- can be random string */, value, { service: key, accessible })

export const getSecureValue: GetSecureValue = async (key) => {
  const result = await Keychain.getGenericPassword({ service: key })
  if (result) {
    return result.password
  }
  return false
}

export const removeSecureValue: RemoveSecureValue = (key) =>
  Keychain.resetGenericPassword({ service: key })
```

2. with internet credentials
```typescript
export const setSecureValue: SetSecureValue = (
  key,
  value,
  accessible = Keychain.ACCESSIBLE.ALWAYS_THIS_DEVICE_ONLY,
) => Keychain.setInternetCredentials(key, key /* <- can be random string */, value, { accessible })

export const getSecureValue: GetSecureValue = async (key) => {
  const result = await Keychain.getInternetCredentials(key)
  if (result) {
    return result.password
  }
  return false
}

export const removeSecureValue: RemoveSecureValue = (key) =>
  Keychain.resetInternetCredentials(key)
```

there is no options, but it shouldn't be a problem to add them

**Comment by pke** [comment]
This should really make it into the README.md

**Comment by jtvargas** [comment]
What would be the difference between saving with `setGenericPassword` and `setInternetCredentials `?

**Comment by fattomhk** [comment]
> What would be the difference between saving with `setGenericPassword` and `setInternetCredentials `?

almost the same.
the `internetCredentials` has more attributes to distinguish the remote access. (i.e. The generic one cannot save the server attribute)
![image](https://user-images.githubusercontent.com/1083049/135819311-d07bd1ac-5bde-4c0e-96b2-ecdbdaa3756f.png)

**Comment by fattomhk** [comment]
> @fattomhk So, if my app use credentials that are also used for a web portal, I need to use `setInternetCredentials` and add the `server` attribute (server will be web portal URL?)

yes, `server` should be the domain name or IP address of url (etc. `github.com`).

**Comment by jtvargas** [comment]
@fattomhk If I use `setInternetCredentials` on iOS and in the key `server`  I use my domain name (etc. `www.exampledomainservervalue.com`), those credentials will appear on my phone keychain (Settings > Passwords) and can be modified via phone settings?

**Comment by fattomhk** [comment]
> @fattomhk If I use `setInternetCredentials` on iOS and in the key `server` I use my domain name (etc. `www.exampledomainservervalue.com`), those credentials will appear on my phone keychain (Settings > Passwords) and can be modified via phone settings?

No. If you want to do that, in short, you need to call `setSharedWebCredentials` 
Normal flow: 
![image](https://user-images.githubusercontent.com/1083049/136132736-ebe34345-6b0b-499a-a5c8-b5f6dcb00415.png)

**Comment by jtvargas** [comment]
> > @fattomhk If I use `setInternetCredentials` on iOS and in the key `server` I use my domain name (etc. `www.exampledomainservervalue.com`), those credentials will appear on my phone keychain (Settings > Passwords) and can be modified via phone settings?
> 
> 
> 
> No. If you want to do that, in short, you need to use `setSharedWebCredentials` and `requestSharedWebCredentials`
> 
> Normal flow: 
> 
> ![image](https://user-images.githubusercontent.com/1083049/136132736-ebe34345-6b0b-499a-a5c8-b5f6dcb00415.png)
> 
> 
> 
> 

But to those functions are only available on iOS, so for Android use “setInternetCredentials”?

**Comment by fattomhk** [comment]
> But to those functions are only available on iOS, so for Android use “setInternetCredentials”?

AFAIK, this lib cannot share pwd to another app or to web. If you want to implement it, maybe smartlock-passwords is the thing you are looking for.

**Comment by rosem** [comment]
Is there an advantage to this over just using setGenericPassword directly? I guess API clarity?
